keps/sig-apps/4443-configurable-pod-failure-policy-reasons/README.md

```diff
@@ -0,0 +1,1000 @@
+
+**Example JobSet for this use case**:
+
+```yaml
+apiVersion: jobset.x-k8s.io/v1alpha2
+kind: JobSet
+metadata:
+  name: restart-job-example
+  annotations:
+    alpha.jobset.sigs.k8s.io/exclusive-topology: {{topologyDomain}} # 1:1 job replica to topology domain assignment
+spec:
+  failurePolicy:
+    rules:
+    # If Job fails due to a pod failing with exit code 2, leave it in a failed state.
+    - action: FailJob
+      targetReplicatedJobs:
+      - simulations
+      onJobFailureReasons:
+      - PodFailurePolicy_ExitCode2  # Job failure reason format: PodFailurePolicy_{ruleName}
+    # If Job fails due to a pod failing with exit code 3, restart that Job.
+    - action: RestartJob
+      targetReplicatedJobs:
+      - simulations
+      onJobFailureReasons:
+      - PodFailurePolicy_ExitCode3  # Job failure reason format: PodFailurePolicy_{ruleName}
+    maxRestarts: 10
+  replicatedJobs:
+  - name: simulations
+    replicas: 10
+    template:
+      spec:
+        parallelism: 1
+        completions: 1
+        backoffLimit: 0
+        # Pod failure policy rules, the names of which are referenced in the JobSet failure policy.
+        podFailurePolicy:
+          rules:
+          - name: ExitCode2
+            action: FailJob
+            onExitCodes:
+              containerName: main
+              operator: In
+              values: [2]
+          - name: ExitCode3
+            action: FailJob
+            onExitCodes:
+              containerName: main
+              operator: In
+              values: [3]
+        template:
+          spec:
+            restartPolicy: Never
+            containers:
+            - name: main
+              image: python:3.10
+              command: ["..."]
+```
+
+### Notes/Constraints/Caveats (Optional)
+
+It should be noted that upon pod failure, the Job's pod failure policy rules
+are evaluated in order, and only the first matching rule is executed, even
+if multiple rules match a pod failure.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+There is a risk to making a field that was previously exclusively managed by the controller,
+to now being configurable by the user. However, as described in the validation section below,
+we are validating against malformed/invalid inputs.
+
+
+## Design Details
+
+#### Defaulting
+
+There will be no defaulting for the new pod failure policy `Name` field.
+When `Name` is unset, the Job controller will set the reason suffix to the index of the rule in the `podFailurePolicy.rules` slice (e.g. `PodFailurePolicy_{index}`).
+
+#### Validation
+- Validate all pod failure policy rule names are unique.
```

Comment:
Can we also validate that the name should not have an integer value of any of the existing indexes? Unless it is its own index?

Reply:
Good point

Reply:
Makes sense - added.